### PR TITLE
[API] Simplify manual pipeline API

### PR DESCRIPTION
**Pull request by merrymercy**
The current manual pipeline API is tedious.

## Before
Require a "start" and an "end" marker for each stage.

```python
class MLP(nn.Module):
    @nn.compact
    def __call__(self, x):
        x = nn.Dense(features=self.hidden_dim * 4)(x)
        alpa.mark_pipeline(name='0', mark_type='end')
        alpa.mark_pipeline(name='1', mark_type='start')
        x = nn.Dense(features=self.hidden_dim * 4)(x)
        return x

@alpa.parallelize
def train_step(state, batch):

    @alpa.manual_layer_construction
    def loss_func(params):
        alpa.mark_pipeline(name='0', mark_type='start')
        out = state.apply_fn(params, batch["x"])
        loss = jnp.mean((out - batch["y"])**2)
        alpa.mark_pipeline(name='1', mark_type='end')
        return loss
```

## After
Only require a `mark_pipeline_boundary` at the stage boundary. No need to add any markers in `loss_func`.

```python
class MLP(nn.Module):

    @nn.compact
    def __call__(self, x):
        x = nn.Dense(features=self.hidden_dim * 4)(x)
        alpa.mark_pipeline_boundary()
        x = nn.Dense(features=self.hidden_dim * 4)(x)
        return x

@alpa.parallelize
def train_step(state, batch):

    @alpa.manual_layer_construction
    def loss_func(params):
        out = state.apply_fn(params, batch["x"])
        loss = jnp.mean((out - batch["y"])**2)
        return loss
```

## Changes
- deprecate `mark_pipeline`
- various improvements to `layer_construction.py`

### CODE REVIEW

**Comment by zhisbug** [comment]
How did you find the loss at the end (and append a marker there)?

**Comment by merrymercy** [comment]
Because the loss function is decorated by `manual_layer_construction`, we can add a marker at the decorated function
